### ISSUE TITLE
Definition of Frontend API Bundle

### ISSUE THREAD

**Issue by dpfaffenbauer** [coreshop/CoreShop#831]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes

In order for CoreShop to be used in PWA/SPA we need a proper Frontend API.

 - Define what API Endpoints we need.

**Comment by davidhoeck** [comment]
## Products and Categories

| Method        | Description | Parameters  |
| ----------- |-------------| --------------|
| __GET__  | All Products | `page` or `skip` and `limit` |
| __GET__  | Single Product by ID       |   `productId` |
| __GET__  | All Categories hierarchic with subcategories      |   `maxDepth` |
| __GET__  | Single Category with its subcategories     |  `categoryId` `maxDepth` |
| __GET__  | All products of a single category      |  `categoryId`, `maxDepth`, `page` or `skip` and `limit` |


## Cart 

| Method        | Description | Parameters  |
| ----------- |-------------| --------------|
| __POST__   | Add Product to the cart  | `productId` `quantity` |
| __PUT__  | Update CartItem  | `cartItemId` `productId` `quantity` |
| __DELETE__ | Remove CartItem from Cart  | `cartItemId` |
| __GET__ | Get the Cart  | `cartIdentifier`(optional) |
| __POST__ | Add a discount/voucher code to the cart | `code` |
| __DELETE__ | Remove a discount/voucher code from the cart | `code` |

## Checkout 
| Method        | Description | Parameters  |
| ----------- |-------------| --------------|
|__GET__ | All Checkout steps  | |
|__PUT__ | Persist a cart step  | `stepIdentifier` |
| __POST__| Create a new address for the customer | - |
| __DELETE__ | Remove an address of the customer |`addressId`|

**Comment by solverat** [comment]
## Cart
- Add multiple items

## Index
- Get Index Data (Data Hub Thing?), related: #692

**Comment by dpfaffenbauer** [comment]
Multiple ways of actually doing this:
 - Custom FrontendAPI Bundle with API Platform
 - Integration with DataHub

@davidhoeck you are working on something like this right?

**Comment by davidhoeck** [comment]
@dpfaffenbauer yes - but still under heavy development and we are trying to find the right architecture. problem is basically that datahub is not built for performance because schema cannot be cached and is initialized on every request

**Comment by dkarlovi** [comment]
@davidhoeck can we fix the "not built for performance" upstream? I'd expect them to cache metadata in prod, same as any other integration.

**Comment by davidhoeck** [comment]
@dkarlovi nope they don't. the schema get's initialized every time. when you take a look at a data hub request with blackfire, you see that the graph (queries + mutations) is built basically every request. I tried to implement a cache layer, but there was a loss of the resolver information.

**Comment by davidhoeck** [comment]
@dpfaffenbauer yep, but we will migrate it to coreshop 3 before that.

**Comment by dkarlovi** [comment]
@davidhoeck that's unfortunate. I'll check it out in Blackfire, thanks for the hint! :beers:

**Comment by dkarlovi** [comment]
Note for the above list, "All products" (and basically any listing) should by default support search, filter and sort params.

**Comment by davidhoeck** [comment]
@dkarlovi well, product listing is handled by the native data hub listing. CoreShop Index is a little bit more complex to use and we did not have implement it yet.

**Comment by dkarlovi** [comment]
We actually have a very custom Elasticsearch based Index which uses the Coreshop interfaces, but implements the search & filter stuff on our own. Not sure this would work by default via the Datahub feature, I'd expect it to be more tightly linked to Pimcore's implementation and database (which here doesn't exist).

 I didn't look into it, but wanted to note listings should expect to accept filters, search and sort related params, for completeness sake.

**Comment by dpfaffenbauer** [comment]
done with headless bundle
